### PR TITLE
Deprecate client_config From Old LXD Client

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This is part of a series of PRs associated with cleaning up the Juju implementation of LXD and implementing clustering.

This removes unnecessary wrappers for dealing with server configuration and replaces them with a new LXD client. Some of the old wrapper methods are no longer implemented, instead relying directly on the upstream library directly.

## QA steps

Modified unit tests are green.
Bootstrapping LXD and deploying to LXD container machines tested and working.

## Documentation changes

No.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Remove tools/lxdclient/client_config.go](https://trello.com/c/5YL9T2w5/199-remove-tools-lxdclient-clientconfiggo)

File: container/lxd/client.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	}
+}
+
+// UpdateServerConfig updates the server configuration with the input values.
+func (c *Client) UpdateServerConfig(cfg map[string]interface{}) error {
+	svr, eTag, err := c.GetServer()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if svr.Config == nil {
```
Comment (howbazaar) [inline container/lxd/client.go]: Are there any concerns here about taking ownership of the map?

Would it make more sense to always create a map?

```go
if srv.Config == nil {
    srv.Config = make(map[string]interface{})
}
for k,v := range cfg {
    srv.Config[k] = v
}
```

File: container/lxd/client.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// UpdateContainerConfig updates the configuration for the container with the
+// input name, using the input values.
+func (c *Client) UpdateContainerConfig(name string, cfg map[string]string) error {
+	container, eTag, err := c.GetContainer(name)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if container.Config == nil {
```
Comment (howbazaar) [inline container/lxd/client.go]: Similarly here.

Comment (howbazaar) [reply container/lxd/client.go]: Interesting that this config map is string to string, but above it is string to interface. Why the inconsistency?

File: container/lxd/client_test.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+package lxd_test
+
+import (
+	"github.com/golang/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/container/lxd"
+	lxdtesting "github.com/juju/juju/container/lxd/testing"
+	coretesting "github.com/juju/juju/testing"
+	"github.com/lxc/lxd/shared/api"
```
Comment (howbazaar) [inline container/lxd/client_test.go]: This import should be in the previous block.

File: container/lxd/client_test.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+func (s *connectionSuite) TestUpdateServerConfig(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+	cSvr := lxdtesting.NewMockContainerServer(ctrl)
+
+	newConfig := map[string]interface{}{"key1": "val1"}
+	updateReq := api.ServerPut{Config: newConfig}
+	gomock.InOrder(
+		cSvr.EXPECT().GetServer().Return(&api.Server{}, eTag, nil),
+		cSvr.EXPECT().UpdateServer(updateReq, eTag).Return(nil),
```
Comment (howbazaar) [inline container/lxd/client_test.go]: I'm still a little unclear on what gomock will assert with respect to this.

Does it assert that the args to UpdateServer match?

File: container/lxd/image.go [hunk 5]
```diff
@@ -17,8 +17,8 @@ import (
 	"github.com/lxc/lxd/shared/api"
 )
 
-// JujuImageServer extends the upstream LXD client.
-type JujuImageServer struct {
+// ImageServer extends the upstream LXD client.
+type ImageServer struct {
```
Comment (howbazaar) [inline container/lxd/image.go]: I approve of this naming change.

File: provider/lxd/environ_raw.go [hunk 6]
```diff
@@ -15,30 +16,37 @@ import (
...
 	lxdStorage
 
 	remote jujulxdclient.Remote
 }
 
+type newClient interface {
+	FindImage(string, string, []lxd.RemoteServer, bool, environs.StatusCallbackFunc) (lxd.SourcedImage, error)
+	GetServer() (server *lxdapi.Server, ETag string, err error)
+	GetConnectionInfo() (info *lxdclient.ConnectionInfo, err error)
+	UpdateServerConfig(map[string]interface{}) error
```
Comment (howbazaar) [inline provider/lxd/environ_raw.go]: The SetServerConfig that is being replaced took both keys and values as strings. Can we be consistent?

File: container/lxd/client_test.go (hunk 4 above)
Comment (manadart) [reply container/lxd/client_test.go]: Yes.

File: container/lxd/client.go (hunk 2 above)
Comment (manadart) [reply container/lxd/client.go]: From a scan of the LXD code, config values can be int/bool/string, but they are validated based on incoming strings. Not sure why it is different between server and container...

File: container/lxd/client.go (hunk 1 above)
Comment (manadart) [reply container/lxd/client.go]: I'll make this change.

File: container/lxd/client_test.go (hunk 3 above)
Comment (manadart) [reply container/lxd/client_test.go]: Acknowledged.

File: provider/lxd/environ_raw.go (hunk 6 above)
Comment (manadart) [reply provider/lxd/environ_raw.go]: See previous comment based on different config types. Should be safe to use strings; I'll change it.

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**